File: luci-app-ipfixprobe/luasrc/model/cbi/ipfixprobe.lua
-- Copyright 2019 Tomas Cejka <cejkat@cesnet.cz>
-- Licensed to the public under the Apache License 2.0.

local sys = require "luci.sys"

m = Map("ipfixprobe", translate("IPFIX exporter"))

s = m:section(TypedSection, "profile", "Export profiles", "List of exporting processes")
s.anonymous = false
s.addremove = true

s:tab("main", "Main options")
s:tab("ipfix", "IPFIX Collector")
s:tab("advanced", "Advanced options")

-- s.template = "cbi/tblsection"

nic = s:taboption("main", DynamicList, "interfaces", translate("Monitored interface"))
nic.rmempty = true
local iface
for k, v in pairs(luci.sys.net.devices()) do
   nic:value(v, v)
end

pl = s:taboption("main", DynamicList, "plugins", translate("List of plugins"), translate("Enable one or more processing plugins. Plugin with parameters must be entered as a custom string, see manual of ipfixprobe for details - https://github.com/CESNET/ipfixprobe/"))
pl:value("basic", "Basic")
pl:value("bstats", "BSTATS (burst packet stats)")
pl:value("dns", "DNS")
pl:value("dnssd", "DNS-Service Discovery")
pl:value("dnssd:txt", "DNS-Service Discovery with extexded TXT")
pl:value("http", "HTTP")
pl:value("idpcontent", "IDPCONTENT (initial data packets content)")
pl:value("netbios", "Netbios")
pl:value("ntp", "NTP")
pl:value("ovpn", "OpenVPN")
pl:value("passivedns", "Passive DNS (dumped DNS data)")
pl:value("phists", "PHISTS (packet histograms - payload sizes and interpacket times)")
pl:value("pstats", "PSTATS (Per-Packet-Information)")
pl:value("quic", "QUIC")
pl:value("rtsp", "RTSP")
pl:value("sip", "SIP")
pl:value("smtp", "SMTP")
pl:value("ssdp", "SSDP")
pl:value("tls", "TLS")
pl:value("wg", "WireGuard")

pl.rmempty = false

splitbiflow = s:taboption("main", Flag, "split_biflow", translate("Split biflow"))
splitbiflow.enabled = "1"
splitbiflow.disabled = "0"
splitbiflow.default = "0"
splitbiflow.rmempty = false
en = s:taboption("main", Flag, "enabled", translate("Enabled"))
en.enabled = "1"
en.disabled = "0"
en.default = "0"
en.rmempty = false

ipfix_host = s:taboption("ipfix", Value, "ipfix_host", translate("IPFIX Collector address"))
ipfix_host.datatype = "host"
ipfix_port = s:taboption("ipfix", Value, "ipfix_port", translate("IPFIX Collector port"))
ipfix_port.datatype = "port"
udp = s:taboption("ipfix", Flag, "ipfix_udp", translate("Use UDP"))
udp.enabled = "1"
udp.disabled = "0"
udp.default = "1"
udp.rmempty = false
link = s:taboption("ipfix", Value, "link", translate("Observation ID"), translate("Identification of link"))
link.default = "1"
link.rmempty = false
dir = s:taboption("ipfix", Value, "dir", translate("Interface"), translate("Identification of interface"))
dir.default = "0"
dir.rmempty = false

raw_blocks = s:taboption("advanced", Value, "raw_blocks", translate("Number of blocks"), translate("Number of allocated blocks for capturing from NIC"))
raw_blocks.default = "3"

raw_packetsinblock = s:taboption("advanced", Value, "raw_packetsinblock", translate("Packets in block"), translate("Number of packets in a block read from NIC"))
raw_packetsinblock.default = "10"

active_timeout = s:taboption("advanced", Value, "active_timeout", translate("Active timeout"), translate("Split long flows into at most N seconds flow records"))
active_timeout.default = "300"

inactive_timeout = s:taboption("advanced", Value, "inactive_timeout", translate("Inactive timeout"), translate("Export a flow record after N seconds of inactivity"))
inactive_timeout.default = "30"

cache_size = s:taboption("advanced", Value, "cache_size", translate("Size of flow cache"), translate("Number of entries given by an exponent 2**N"))
cache_size.default = "10"

cache_line = s:taboption("advanced", Value, "cache_line", translate("Size of flow cache line"), translate("Number of entries given by an exponent 2**N"))
cache_line.default = "2"

respawn = s:taboption("advanced", Flag, "respawn", translate("Respawn"))
respawn.enabled = "1"
respawn.disabled = "0"
respawn.default = "1"

respawn_threshold = s:taboption("advanced", Value, "respawn_threshold", translate("Respawn threshold"))
respawn_threshold.default = "3600"

respawn_timeout = s:taboption("advanced", Value, "respawn_timeout", translate("Respawn timeout"))
respawn_timeout.default = "5"

respawn_retry = s:taboption("advanced", Value, "respawn_retry", translate("Respawn retry"))
respawn_retry.default = "5"

core = s:taboption("advanced", Flag, "core", translate("Core dumps"))
core.enabled = "1"
core.disabled = "0"
core.default = "0"

m.on_after_commit = function()
  luci.sys.call('/etc/init.d/ipfixprobe reload')
  luci.sys.call('/etc/init.d/ipfixprobe restart')
end

return m

File: luci-app-ipfixprobe/htdocs/luci-static/resources/view/services/ipfixprobe.js
'use strict';
'require ui';
'require uci';
'require view';
'require form';
'require validation';
'require tools.widgets as widgets';

function validatePowerOf2(sid, s) {
	if (s == null || s == '')
		return true;

	var x = validation.parseInteger(s);

	if (isNaN(x) || x < 1) {
		// Must be unsigned integer
		return _('Expecting: positive integer value');
	}

	if (Math.log2(x) % 1 !== 0) {
		// Must be power of 2
		return _('Expecting: power of 2, e.g. 16, 128, 2048,...');
	}

	return true;
}

return view.extend({
	render: function(data) {
		var m, s, o;

		m = new form.Map('ipfixprobe', _('IPFIX exporter'));

		s = m.section(form.TypedSection, 'profile', 'Export profiles', 'List of exporting processes');
		s.anonymous = false;
		s.addremove = true;

		s.tab('main', 'Main options');
		s.tab('ipfix', 'IPFIX Collector');
		s.tab('advanced', 'Advanced options');

		o = s.taboption('main', form.Flag, 'enabled', _('Enabled'));
		o.default = '0';
		o.rmempty = false;

		o = s.taboption('main', widgets.DeviceSelect, 'interfaces', _('Monitored interface'));
		o.multiple = true;
		o.noaliases = true;
		o.rmempty = false;

		o = s.taboption('main', form.DynamicList, 'plugins', _('List of plugins'),
			_('Enable one or more processing plugins. Plugin with parameters must be entered as a custom string, see manual of ipfixprobe for details - <a href="https://github.com/CESNET/ipfixprobe/" target="_blank">https://github.com/CESNET/ipfixprobe/</a>'));
		o.rmempty = false;
		o.value('basic', 'Basic');
		o.value('bstats', 'BSTATS (burst packet stats)');
		o.value('dns', 'DNS');
		o.value('dnssd', 'DNS-Service Discovery');
		o.value('dnssd;txt', 'DNS-Service Discovery with extexded TXT');
		o.value('http', 'HTTP');
		o.value('idpcontent', 'IDPCONTENT (initial data packets content)');
		o.value('netbios', 'Netbios');
		o.value('ntp', 'NTP');
		o.value('ovpn', 'OpenVPN');
		o.value('passivedns', 'Passive DNS (dumped DNS data)');
		o.value('phists', 'PHISTS (packet histograms - payload sizes and interpacket times)');
		o.value('pstats', 'PSTATS (Per-Packet-Information)');
		o.value('quic', 'QUIC');
		o.value('rtsp', 'RTSP');
		o.value('sip', 'SIP');
		o.value('smtp', 'SMTP');
		o.value('ssdp', 'SSDP');
		o.value('tls', 'TLS');
		o.value('wg', 'WireGuard');

		o = s.taboption('main', form.Flag, 'split_biflow', _('Split biflow'));
		o.default = '0';
		o.rmempty = false;

		o = s.taboption('ipfix', form.Value, 'ipfix_host', _('IPFIX Collector address'));
		o.datatype = 'host';

		o = s.taboption('ipfix', form.Value, 'ipfix_port', _('IPFIX Collector port'));
		o.datatype = 'port';

		o = s.taboption('ipfix', form.Value, 'ipfix_mtu', _('IPFIX Collector MTU'));
		o.datatype = 'uinteger';
		o.default = '1452';
		o.rmempty = false;

		o = s.taboption('ipfix', form.Flag, 'ipfix_udp', _('Use UDP'));
		o.default = '1';
		o.rmempty = false;

		o = s.taboption('ipfix', form.Value, 'link', _('Observation ID'), _('Identification of link'));
		o.datatype = 'uinteger';
		o.default = '1';
		o.rmempty = false;

		o = s.taboption('ipfix', form.Value, 'dir', _('Interface'), _('Identification of interface'));
		o.datatype = 'uinteger';
		o.default = '0';
		o.rmempty = false;

		o = s.taboption('advanced', form.Value, 'active_timeout', _('Active timeout'),
			_('Split long flows into at most N seconds flow records.'));
		o.datatype = 'uinteger';
		o.default = '300';

		o = s.taboption('advanced', form.Value, 'inactive_timeout', _('Inactive timeout'),
			_('Export a flow record after N seconds of inactivity.'));
		o.datatype = 'uinteger';
		o.default = '30';

		o = s.taboption('advanced', form.Value, 'raw_blocks', _('Number of blocks'),
			_('Number of allocated blocks for capturing from NIC.'));
		o.datatype = 'uinteger';
		o.default = '2';

		o = s.taboption('advanced', form.Value, 'raw_packetsinblock', _('Packets in block'),
			_('Number of packets in a block read from NIC.'));
		o.datatype = 'uinteger';
		o.default = '10';

		o = s.taboption('advanced', form.Value, 'cache_size', _('Size of flow cache'),
			_('Number of entries. Must be power of 2 (i.e. 2**N).'));
		o.default = '1024';
		o.validate = validatePowerOf2;

		o = s.taboption('advanced', form.Value, 'cache_line', _('Size of flow cache line'),
			_('Number of entries. Must be power of 2 (i.e. 2**N).'));
		o.default = '4';
		o.validate = validatePowerOf2;

		o = s.taboption('advanced', form.Flag, 'respawn', _('Respawn'),
			_('Enable respawn of crashed process.'));
		o.default = '1';

		o = s.taboption('advanced', form.Value, 'respawn_threshold', _('Respawn threshold'),
			_('Timeout in seconds for restarting a service after it closes.'));
		o.datatype = 'uinteger';
		o.default = '3600';

		o = s.taboption('advanced', form.Value, 'respawn_timeout', _('Respawn timeout'),
			_('Maximum time in seconds to wait for a process respawn to complete.'));
		o.datatype = 'uinteger';
		o.default = '5';

		o = s.taboption('advanced', form.Value, 'respawn_retry', _('Respawn retry'),
			_('Maximum number of times to attempt respawning before giving up, 0 means never stop trying to respawn.'));
		o.datatype = 'uinteger';
		o.default = '5';

		o = s.taboption('advanced', form.Flag, 'core', _('Core dumps'));
		o.default = '0';

		return m.render();
	}
})
